x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/AuthenticationContext.java

```diff
@@ -48,6 +48,13 @@ public Subject getEffectiveSubject() {
         return effectiveSubject;
     }
 
+    /**
+     * Authenticated with an API key and no run-as
+     */
+    public boolean isApiKey() {
+        return AuthenticationType.API_KEY.equals(type) && false == isRunAs();
+    }
+
```

Comment:
I'd suggest we move away from the AuthenticationType checking in these classes as much as possible. One of the main reasons for adding `AuthenticationContext` and `Subject` is to make it possible to answer these sorts of question without cross checking information from different places so that we can rely on `Subject` most of the times, e.g, the above can be:
```
getEffectiveSubject().getType() == Subject.Type.API_KEY
```

Reply:
I have followed your suggestion, BUT it changes the behavior of the check.
The question is whether the tokens created by an API Key ought to have access to the resources created by the API Key.
The existing code prevents it, but by analogy with the realm-based behavior, it should.

What do you think?

Reply:
Yes an API key and tokens created by it should be considered as equivalent when it comes to resource sharing. You are right that the existing code does not cater for it. But existing code (8.0 and earlier) also does not work properly for creating tokens using API key. The tokens created by API keys always have no privilege. So it is arguable that they are not considered as equivalent in existing code. Since #80926 (8.1), token creation with API keys is properly supported and now is the right time to also properly handle the resource sharing between them.

Reply:
I agree, thanks for the input.